<a href="https://colab.research.google.com/github/adalbertii/Modele-klasyfikacyjne/blob/main/Embedding_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
import numpy as np

In [2]:
# utworzenie prostego modelu z warstwą Embedding
model = Sequential()
embedding_layer = Embedding(input_dim=10,output_dim=4,input_length=2)
model.add(embedding_layer)
model.compile('adam','mse')

# input_dim - rozmiar słownika
# output_dim - długość wektora dla  każdego słowa
# input_length - maksymalna długość sekwencji

# W powyższym przykładzie ustawiamy 10 jako rozmiar słownictwa, ponieważ będziemy kodować liczby od 0 do 9.
# Chcemy, aby długość wektora słów wynosiła 4, stąd output_dim jest ustawione na 4.
# Długość sekwencji wejściowej do warstwy osadzania będzie wynosić 2

In [3]:
# Teraz przekażmy przykładowe dane wejściowe do naszego modelu i zobaczmy wyniki.
input_data = np.array([[1,2]])
pred = model.predict(input_data)
print(input_data.shape)
print(pred)

1/1 [==============================] - 0s 145ms/step
(1, 2)
[[[-0.02334188  0.03078547  0.01017154  0.04552039]
  [ 0.00770851 -0.01431287  0.01345507  0.005794  ]]]


Jak widać, każde słowo (1 i 2) jest reprezentowane przez wektor o długości 4. Jeśli wydrukujemy wagi warstwy osadzania, otrzymamy poniższy wynik.

In [7]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 2, 4)              40        
                                                                 
Total params: 40 (160.00 Byte)
Trainable params: 40 (160.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [6]:
print(model.weights)

[<tf.Variable 'embedding/embeddings:0' shape=(10, 4) dtype=float32, numpy=
array([[ 0.02532529,  0.02959583, -0.02077184,  0.03307004],
       [-0.02334188,  0.03078547,  0.01017154,  0.04552039],
       [ 0.00770851, -0.01431287,  0.01345507,  0.005794  ],
       [-0.04999755,  0.01052343,  0.035213  , -0.03954636],
       [ 0.03153438,  0.0402245 ,  0.04568147, -0.03027884],
       [ 0.01681492, -0.01277689, -0.02017144,  0.00512385],
       [-0.01597724, -0.00012515, -0.02283381, -0.02160322],
       [ 0.04674724,  0.0115404 , -0.02743992,  0.00157394],
       [ 0.00702375, -0.00607336,  0.00393804, -0.04824213],
       [-0.02383035,  0.01415158,  0.03827745, -0.02038814]],
      dtype=float32)>]


Te wagi są w zasadzie reprezentacjami wektorowymi słów w słownictwie. Jak omówiliśmy wcześniej, jest to tabela odnośników o rozmiarze 10 x 4, dla słów od 0 do 9. Pierwsze słowo (0) jest reprezentowane przez pierwszy wiersz w tej tabeli, czyli

[0.02532529,  0.02959583, -0.02077184,  0.03307004]

W tym przykładzie nie wytrenowaliśmy warstwy osadzania. Wagi przypisane do wektorów słów są inicjowane losowo.